include/deal.II/lac/tensor_product_matrix.h

```diff
@@ -311,96 +691,193 @@ template <int dim, typename Number, int size>
+  std::vector<Number> deriv_matrix_flat ;
+  std::vector<Number> eigenvalues_flat ;
+  std::vector<Number> eigenvectors_flat ;
+  mass_matrix_flat.reserve (nm_flat_size) ;
+  deriv_matrix_flat.reserve (nm_flat_size) ;
+  eigenvalues_flat.reserve (n_flat_size) ;
+  eigenvectors_flat.reserve (nm_flat_size) ;
+  std::array<unsigned int,macro_size> offsets_nm ;
+  std::array<unsigned int,macro_size> offsets_n ;
+  for (int dir = 0; dir < dim; ++dir)
     {
-      eval.template apply<0, true, false>(M, src, t+n);
-      eval.template apply<1, true, false>(M, t+n, t);
-      eval.template apply<2, true, false>(A, t, dst);
-      eval.template apply<1, true, false>(A, t+n, t);
-      eval.template apply<0, true, false>(A, src, t+n);
-      eval.template apply<1, true, true> (M, t+n, t);
-      eval.template apply<2, true, true> (M, t, dst);
+      Assert (size == -1 ||
+              (size > 0 && static_cast<unsigned int>(size) == mass_matrix[dir].n_rows()),
+              ExcDimensionMismatch(size, mass_matrix[dir].n_rows()));
+      AssertDimension (mass_matrix[dir].n_rows(), mass_matrix[dir].n_cols());
+      AssertDimension (mass_matrix[dir].n_rows(), derivative_matrix[dir].n_rows());
+      AssertDimension (mass_matrix[dir].n_rows(), derivative_matrix[dir].n_cols());
+
+      const unsigned int n_rows = mass_matrix[dir].n_rows() ;
+      const unsigned int n_cols = mass_matrix[dir].n_cols() ;
+      const unsigned int nm = n_rows * n_cols ;
+
+      mass_matrix_flat.resize (macro_size*nm) ;
+      deriv_matrix_flat.resize (macro_size*nm) ;
+      eigenvalues_flat.resize (macro_size*n_rows) ;
+      eigenvectors_flat.resize (macro_size*nm) ;
+      for (unsigned int vv=0; vv<macro_size; ++vv)
+        offsets_nm[vv] = nm * vv ;
+      for (unsigned int vv=0; vv<macro_size; ++vv)
+        offsets_n[vv] = n_rows * vv ;
+
+      vectorized_transpose_and_store (false, nm, &(mass_matrix[dir](0,0))
+                                      , offsets_nm.cbegin(), mass_matrix_flat.data()) ;
+      vectorized_transpose_and_store (false, nm, &(derivative_matrix[dir](0,0))
+                                      , offsets_nm.cbegin(), deriv_matrix_flat.data()) ;
+
+      const Number *mass_cbegin = mass_matrix_flat.data() ;
+      const Number *deriv_cbegin = deriv_matrix_flat.data() ;
+      Number *eigenvec_begin = eigenvectors_flat.data() ;
+      Number *eigenval_begin = eigenvalues_flat.data() ;
+
+      spectral_assembly<Number> (mass_cbegin, deriv_cbegin, n_rows, n_cols
+                                 , eigenval_begin, eigenvec_begin) ;
+      for (unsigned int lane = 1; lane < macro_size; ++lane)
+        {
+          std::advance (mass_cbegin, nm) ;
+          std::advance (deriv_cbegin, nm) ;
+          std::advance (eigenvec_begin, nm) ;
+          std::advance (eigenval_begin, n_rows) ;
+          spectral_assembly<Number> (mass_cbegin, deriv_cbegin, n_rows, n_cols
+                                     , eigenval_begin, eigenvec_begin) ;
```

Comment:
I guess here you are overusing the standard template library a bit and hide the actual intent. Why not simply put `mass_cbegin+nm*lane, deriv_cbegin+nm*lane` etc as the arguments? This way you also only have to spell `spectral_assembly` once rather than twice.

Reply:
Yes, you are right, this is done overly complicated. I will fix it.